### PR TITLE
Switch from libsoup to libcurl

### PR DESCRIPTION
**Pull request by hughsie**
The former drags on glib-networking and then gsettings-desktop-schemas, which
add over 5Mb to the minimal IoT and CoreOS composes. Everything already uses
libcurl (even NetworkManager!) and so this is an easy way to reduce image size.

### CODE REVIEW

File: libfwupd/fwupd-client.c
Comment (lucab) [inline libfwupd/fwupd-client.c:94]: Should this early-return if `helper == NULL`, mimicking `g_free` behavior?

File: libfwupd/fwupd-client.c
Comment (lucab) [inline libfwupd/fwupd-client.c:313]: `curl_off_t` is a signed type; perhaps it should have some non-negative clamping logic before casting to `guint`? Also, should division-by-0 a concern on the `dltotal` here?

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:94]: I believe G_DEFINE_AUTOPTR_CLEANUP_FUNC does that for us.

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:313]: Hmm, very valid. Also, we use the same function for uploads but don't look at ultotal at all.

Comment (hughsie) [reply libfwupd/fwupd-client.c:313]: @lucab how about something like this?

    /* calculate percentage */
    if (dltotal >= 0 && dlnow >= 0 && dlnow <= dltotal) {
        guint percentage = (guint) ((100 * dlnow) / dltotal);
        g_debug ("download progress: %u%%", percentage);
        fwupd_client_set_percentage (self, percentage);
    } else if (ultotal >= 0 && ulnow >= 0 && ulnow <= ultotal) {
        guint percentage = (guint) ((100 * ulnow) / ultotal);
        g_debug ("upload progress: %u%%", percentage);
        fwupd_client_set_percentage (self, percentage);
    }

File: libfwupd/fwupd-client.c
Comment (lucab) [reply libfwupd/fwupd-client.c:94]: Absolutely true, I should have double-checked the docs first, sorry for the noise.

Comment (hughsie) [reply libfwupd/fwupd-client.c:94]: Not noise! I really appreciate the sanity checks!

File: libfwupd/fwupd-client.c
Comment (lucab) [reply libfwupd/fwupd-client.c:313]: :+1: but I think you'd want a `> 0` on both `*total` conditions (instead of the `>= 0`).

Comment (hughsie) [reply libfwupd/fwupd-client.c:313]: Fixed, ta.

File: snap/snapcraft.yaml
Comment (superm1) [inline snap/snapcraft.yaml:11]: I don't think it can go to core20 yet. I'll check out what is going on with snap. Do you know if the curl version you prescribed is actually required, or can it go older?

File: meson.build
Comment (superm1) [inline meson.build:211]: Can it be older or is this just the first version you tried?

Comment (hughsie) [reply meson.build:211]: It needs to be that for the CURLU support.

File: snap/snapcraft.yaml
Comment (hughsie) [reply snap/snapcraft.yaml:11]: I'll admit I don't know the nuances of the snap base snap -- it's all a bit confusing.

**Comment by superm1** [comment]
well this is unfortunately where the approach of split CI between build container and build job fails..  Adding new dependencies.  Will have a thought what can be done other than manually fixing the containers.

**Comment by superm1** [comment]
> The former drags on glib-networking and then gsettings-desktop-schemas, which
add over 5Mb to the minimal IoT and CoreOS composes. Everything already uses
libcurl (even NetworkManager!) and so this is an easy way to reduce image size.

As I can tell, gusb is also pulling in libsoup.

File: snap/snapcraft.yaml
Comment (superm1) [reply snap/snapcraft.yaml]: https://snapcraft.io/docs/base-snaps explains it a bit

**Review by superm1 — APPROVED** [review]
squash up my CI fixes and then LGTM.  We can also drop the libsoup in `dependencies.xml` after PR is merged (due to ABI check).

**Comment by hughsie** [comment]
> squash up my CI fixes and then LGTM

Thanks for your help with this; I'd like to finish the gnome-software plugin work first -- we actually grab the SoupSession (using `g_object_get`, yuk!) in later gnome-software versions and I can use that as an excuse to port the plugin to the new async stuff we have.

**Comment by hughsie** [comment]
If it's interesting, https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/538

**Comment by superm1** [comment]
> If it's interesting, https://gitlab.gnome.org/GNOME/gnome-software/-/merge_requests/538

Hmm so that means that moving to fwupd 1.5.2 and using "old" GS is off the table?

**Comment by hughsie** [comment]
> old GS is off the table

I think so; it wasn't my finest moment of forward planning to use the GObject property like that. I think backporting the fwupd plugin changes will be fine, but we might want to hold off the libcurl stuff until 1.6.x even.

**Comment by superm1** [comment]
As a thought, what about offering a compilation option of which one to use, with the intention that eventually soup is dropped?

**Comment by superm1** [comment]
Or maybe use both and to "deprecate" the soup one for an ABI change (which really it sounds like it might be).

**Comment by hughsie** [comment]
> Or maybe use both

The l33t hacker in me wants to dlopen the libsoup.so at runtime and just create the object that way, so there's no explicit library dep. It's kinda sucky the very same binary fwupd.rpm package gets used on coreos and the Fedora Workstation.

**Comment by superm1** [comment]
> > Or maybe use both
> 
> The l33t hacker in me wants to dlopen the libsoup.so at runtime and just create the object that way, so there's no explicit library dep. It's kinda sucky the very same binary fwupd.rpm package gets used on coreos and the Fedora Workstation.

I guess I go back and forth - besides GS does any other fwupd client steal the soup session?  If not, as long as gnome-software "cleanly" aborts/blows up let's not make a bunch of noise about the small ABI break.  We can make sure that a new gnome-software and fwupd 1.5.2 flow into distros around the same time.

**Comment by superm1** [comment]
The only other one that came to mind to maybe be problematic is the snap store (gnome-software fork IIRC?).  But they carry the fwupd library as a build time dependency.  So I'd think when they move up to 1.5.2 in their snap they can update their fwupd support at the same time.  CC @kenvandine FYI ^

**Comment by kenvandine** [comment]
> The only other one that came to mind to maybe be problematic is the snap store (gnome-software fork IIRC?). But they carry the fwupd library as a build time dependency. So I'd think when they move up to 1.5.2 in their snap they can update their fwupd support at the same time. CC @kenvandine FYI ^

Thanks for the heads up!

**Comment by superm1** [comment]
> If not, as long as gnome-software "cleanly" aborts/blows up

Running G-S master against this branch I don't see a functional issue.  I guess the user agent is going to be wrong, and if you had a proxy set it might not work properly.

**Comment by hughsie** [comment]
> Running G-S master against this branch I don't see a functional issue

Does firmware still get downloaded tho? I might try the dlopen thing myself tomorrow and see if we can have the best of both worlds, i.e. dropping the dep and not breaking internal ABI.

**Comment by superm1** [comment]
> Does firmware still get downloaded tho? I might try the dlopen thing myself tomorrow and see if we can have the best of both worlds, i.e. dropping the dep and not breaking internal ABI.

I only tried refreshing metadata since I didn't have something pending an update at the moment.

**Comment by hughsie** [comment]
Answering my own question:

    14:05:56:0937 libsoup soup_session_send_message: assertion 'SOUP_IS_SESSION (session)' failed

I'm working on something horrific.

**Comment by hughsie** [comment]
@superm1 try to review the above commit without being a little bit sick in your mouth.

File: libfwupd/fwupd-client.c
Comment (superm1) [inline libfwupd/fwupd-client.c:83]: this comment is wrong now..

File: libfwupd/fwupd-client.c
Comment (superm1) [inline libfwupd/fwupd-client.c:4280]: hardcoded paths?  Is there a cleaner way to do this?

File: libfwupd/fwupd-client.c
Comment (superm1) [inline libfwupd/fwupd-client.c:4294]: maybe unload the module in this error path?

File: libfwupd/fwupd-client.c
Comment (superm1) [inline libfwupd/fwupd-client.c:4282]: hardcoding this extension won't work in !linux

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:83]: Fixed.

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:4294]: I thought I'd leave it be, otherwise we'll keep trying to open libsoup for every property read.

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:4282]: Valid, but I don't think it matters; this is just for gnome-software.

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:4280]: I originally just had FWUPD_LIBDIR but that's prefixed with the, well, prefix and then it breaks for when you're developing fwupd. Perhaps that's actually okay, and we just use FWUPD_LIBDIR which will be the right thing when installed systemwide.

File: libfwupd/fwupd-client.c
Comment (superm1) [reply libfwupd/fwupd-client.c:4294]: But don't you want to do to that?  What if someone didn't have libsoup installed and then installed it later at runtime?

File: libfwupd/fwupd-client.c
Comment (superm1) [reply libfwupd/fwupd-client.c:4282]: Maybe don't even bother compiling the code for mingw builds?

File: libfwupd/fwupd-client.c
Comment (superm1) [reply libfwupd/fwupd-client.c:4280]: What I was thinking was a `fu_common_get_path` type of solution, which will let this work well in the snap too, and also for developing fwupd.

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:4294]: Possible indeed. Fixed.

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:4282]: Yes, I'd feel a lot happier if all this evil was behind some kind of ifdef. Will fix.

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:4280]: fu_common_get_path is libfwupdplugin, not libfwupd  :(

Comment (superm1) [reply libfwupd/fwupd-client.c:4280]: ah..  Well in that case I guess yeah just make `FWUPD_LIBDIR` only.  Need to keep reminding myself it's just for gnome-software support, not general.  Snap-store will just need move at the same time they upgrade their library.

And `libfwupd3` can't come soon enough :)

Comment (hughsie) [reply libfwupd/fwupd-client.c:4280]: I actually think this would be fine for snap:

		tmp = g_getenv ("SNAP");
		if (tmp != NULL) {
			g_ptr_array_add (libdirs,
					 g_build_filename (tmp, FWUPD_LIBDIR, NULL));
		}

Comment (hughsie) [reply libfwupd/fwupd-client.c:4280]: although, maybe that's not required at all if ubuntu-software, ahem, gnome-software is using the system libfwupd and not the snapped version?

Comment (hughsie) [reply libfwupd/fwupd-client.c:4280]: I do like the simplicity of the FWUPD_LIBDIR-only approach -- it's just fixing the one problem

Comment (superm1) [reply libfwupd/fwupd-client.c:4280]: The snap store is fully confined and talks to the host daemon.  So expect the client library to move with the confined store itself.
@kenvandine am I right that snap store can just update interface at the same time the library version bumps?

Comment (hughsie) [reply libfwupd/fwupd-client.c:4280]: >  So expect the client library to move with the confined store itself

That seems sane; I think we don't need the snap-specific path then.

Comment (hughsie) [reply libfwupd/fwupd-client.c]: I've made it a lot simpler, we can dig this back up if indeed snap needs the fix.

Comment (kenvandine) [reply libfwupd/fwupd-client.c]: > The snap store is fully confined and talks to the host daemon. So expect the client library to move with the confined store itself.
> @kenvandine am I right that snap store can just update interface at the same time the library version bumps?

Yes, we build libfwupd into the snap and talk to the host service.  So no dependency on the lib on the host.  So FWUPD_LIBDIR should be fine.

File: libfwupd/fwupd-client.c
Comment (superm1) [inline libfwupd/fwupd-client.c:4525]: isn't this also ABI breaking?  I guess _gnome-software didn't actually use it though, so don't make too much noise_?

**Review by superm1 — APPROVED** [review]
I already look forward to our ABI bump to drop this soup emulation code.  I think in the distro packages we may consider to change the defaults to not compile it once we get gnome-software new stuff landed in the big ones.

File: libfwupd/fwupd-client.c
Comment (hughsie) [reply libfwupd/fwupd-client.c:4525]: > so don't make too much noise

Right! Sssh! :)

**Pull request merged**
